server/snapshot/snapshot_test.go

```diff
@@ -0,0 +1,209 @@
+package snapshot
+
+import (
+	"bytes"
+	"encoding/json"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/docker/notary/server/storage"
+	"github.com/docker/notary/tuf/data"
+	"github.com/docker/notary/tuf/signed"
+)
+
+func TestSnapshotExpired(t *testing.T) {
+	sn := &data.SignedSnapshot{
+		Signatures: nil,
+		Signed: data.Snapshot{
+			Expires: time.Now().AddDate(-1, 0, 0),
+		},
+	}
+	assert.True(t, snapshotExpired(sn), "Snapshot should have expired")
+}
+
+func TestSnapshotNotExpired(t *testing.T) {
+	sn := &data.SignedSnapshot{
+		Signatures: nil,
+		Signed: data.Snapshot{
+			Expires: time.Now().AddDate(1, 0, 0),
+		},
+	}
+	assert.False(t, snapshotExpired(sn), "Snapshot should NOT have expired")
+}
+
+func TestGetSnapshotKeyCreate(t *testing.T) {
+	store := storage.NewMemStorage()
+	crypto := signed.NewEd25519()
+	k, err := GetOrCreateSnapshotKey("gun", store, crypto, data.ED25519Key)
+	assert.Nil(t, err, "Expected nil error")
+	assert.NotNil(t, k, "Key should not be nil")
+
+	k2, err := GetOrCreateSnapshotKey("gun", store, crypto, data.ED25519Key)
+
+	assert.Nil(t, err, "Expected nil error")
+
+	// trying to get the same key again should return the same value
+	assert.Equal(t, k, k2, "Did not receive same key when attempting to recreate.")
+	assert.NotNil(t, k2, "Key should not be nil")
+}
+
+func TestGetSnapshotKeyExisting(t *testing.T) {
+	store := storage.NewMemStorage()
+	crypto := signed.NewEd25519()
+	key, err := crypto.Create(data.CanonicalSnapshotRole, data.ED25519Key)
+	assert.NoError(t, err)
+
+	store.SetKey("gun", data.CanonicalSnapshotRole, data.ED25519Key, key.Public())
+
+	k, err := GetOrCreateSnapshotKey("gun", store, crypto, data.ED25519Key)
+	assert.Nil(t, err, "Expected nil error")
+	assert.NotNil(t, k, "Key should not be nil")
+	assert.Equal(t, key, k, "Did not receive same key when attempting to recreate.")
+	assert.NotNil(t, k, "Key should not be nil")
+
+	k2, err := GetOrCreateSnapshotKey("gun", store, crypto, data.ED25519Key)
+
+	assert.Nil(t, err, "Expected nil error")
+
+	// trying to get the same key again should return the same value
+	assert.Equal(t, k, k2, "Did not receive same key when attempting to recreate.")
+	assert.NotNil(t, k2, "Key should not be nil")
+}
+
+type keyStore struct {
+	getCalled bool
+	k         data.PublicKey
+}
+
+func (ks *keyStore) GetKey(gun, role string) (string, []byte, error) {
+	defer func() { ks.getCalled = true }()
+	if ks.getCalled {
+		return ks.k.Algorithm(), ks.k.Public(), nil
+	}
+	return "", nil, &storage.ErrNoKey{}
+}
+
+func (ks keyStore) SetKey(gun, role, algorithm string, public []byte) error {
+	return &storage.ErrKeyExists{}
+}
+
+func TestGetSnapshotKeyExistsOnSet(t *testing.T) {
```

Comment:
Yeah, I did a tiny refactor on `storage.MetaStore` to create a tiny `KeyStore` interface I could implement to let me hit all the errors in `GetOrCreateKey`. Once the snapshot and timestamp versions are unified it'll be great.